src/coprocessor/dag/rpn_expr/types.rs

```diff
@@ -0,0 +1,803 @@
+        let mut expr_nodes = Vec::new();
+        Self::append_rpn_nodes_recursively(
+            def,
+            time_zone,
+            &mut expr_nodes,
+            super::map_pb_sig_to_rpn_func,
+        )?;
+        Ok(RpnExpressionNodeVec(expr_nodes))
+    }
+
+    /// Transforms eval tree nodes into RPN nodes.
+    ///
+    /// Suppose that we have a function call:
+    ///
+    /// ```ignore
+    /// A(B, C(E, F, G), D)
+    /// ```
+    ///
+    /// The eval tree looks like:
+    ///
+    /// ```ignore
+    ///           +---+
+    ///           | A |
+    ///           +---+
+    ///             |
+    ///   +-------------------+
+    ///   |         |         |
+    /// +---+     +---+     +---+
+    /// | B |     | C |     | D |
+    /// +---+     +---+     +---+
+    ///             |
+    ///      +-------------+
+    ///      |      |      |
+    ///    +---+  +---+  +---+
+    ///    | E |  | F |  | G |
+    ///    +---+  +---+  +---+
+    /// ```
+    ///
+    /// We need to transform the tree into RPN nodes:
+    ///
+    /// ```ignore
+    /// B E F G C D A
+    /// ```
+    ///
+    /// The transform process is very much like a post-order traversal. This function does it
+    /// recursively.
+    fn append_rpn_nodes_recursively<F>(
+        mut def: Expr,
+        time_zone: Tz,
+        rpn_nodes: &mut Vec<RpnExpressionNode>,
+        fn_mapper: F,
+    ) -> Result<()>
+    where
+        F: Fn(tipb::expression::ScalarFuncSig) -> Result<Box<dyn RpnFunction>> + Copy,
+    {
+        // TODO: We should check whether node types match the function signature. Otherwise there
+        // will be panics when the expression is evaluated.
+
+        use crate::coprocessor::codec::mysql::{Decimal, Duration, Json, Time, MAX_FSP};
+        use crate::util::codec::number;
+        use std::convert::{TryFrom, TryInto};
+
+        #[inline]
+        fn handle_expr_tree_node_null(
+            field_type: FieldType,
+            eval_type: EvalType,
+            rpn_nodes: &mut Vec<RpnExpressionNode>,
+        ) -> Result<()> {
+            let scalar_value = match eval_type {
+                EvalType::Int => ScalarValue::Int(None),
+                EvalType::Real => ScalarValue::Real(None),
+                EvalType::Decimal => ScalarValue::Decimal(None),
+                EvalType::Bytes => ScalarValue::Bytes(None),
+                EvalType::DateTime => ScalarValue::DateTime(None),
+                EvalType::Duration => ScalarValue::Duration(None),
+                EvalType::Json => ScalarValue::Json(None),
+            };
+            rpn_nodes.push(RpnExpressionNode::Constant {
+                value: scalar_value,
+                field_type,
+            });
+            Ok(())
+        }
+
+        #[inline]
+        fn handle_expr_tree_node_int64(
+            field_type: FieldType,
+            eval_type: EvalType,
+            rpn_nodes: &mut Vec<RpnExpressionNode>,
+            def_without_field_type: Expr,
```

Comment:
I think the prefix `def` in `def_without_field_type` is not clear.

Reply:
tree_node_without_field_type?